pkg/costmodel/costmodel.go

```diff
@@ -828,32 +848,62 @@ func findDeletedNodeInfo(cli prometheusClient.Client, missingNodes map[string]*c
 	return nil
 }
 
-func getContainerAllocation(req []*util.Vector, used []*util.Vector, allocationType string) []*util.Vector {
-	// The result of the normalize operation will be a new []*util.Vector to replace the requests
-	allocationOp := func(r *util.Vector, x *float64, y *float64) bool {
-		if x != nil && y != nil {
-			x1 := *x
-			if math.IsNaN(x1) {
-				log.Warnf("NaN value found during %s allocation calculation for requests.", allocationType)
-				x1 = 0.0
-			}
-			y1 := *y
-			if math.IsNaN(y1) {
-				log.Warnf("NaN value found during %s allocation calculation for used.", allocationType)
-				y1 = 0.0
-			}
-
-			r.Value = math.Max(x1, y1)
-		} else if x != nil {
-			r.Value = *x
-		} else if y != nil {
-			r.Value = *y
+// getContainerAllocation takes the max between request and usage. This function
+// returns a slice containing a single element describing the container's
+// allocation.
+//
+// Additionally, the timestamp of the allocation will be the highest value
+// timestamp between the two vectors. This mitigates situations where
+// Timestamp=0. This should have no effect on the metrics emitted by the
+// CostModelMetricsEmitter
+func getContainerAllocation(req *util.Vector, used *util.Vector, allocationType string) []*util.Vector {
+	var result []*util.Vector
+
+	if req != nil && used != nil {
+		x1 := req.Value
+		if math.IsNaN(x1) {
+			log.Warnf("NaN value found during %s allocation calculation for requests.", allocationType)
+			x1 = 0.0
+		}
+		y1 := used.Value
+		if math.IsNaN(y1) {
+			log.Warnf("NaN value found during %s allocation calculation for used.", allocationType)
+			y1 = 0.0
+		}
+		result = []*util.Vector{
+			{
+				Value:     math.Max(x1, y1),
+				Timestamp: math.Max(req.Timestamp, used.Timestamp),
```

Comment:
@nikovacevic Curious to hear what you think about doing `math.Max(req.Timestamp, used.Timestamp)`. I wanted to remove the assumption that the caller would set `req.Timestamp = float64(time.Now().UTC().Unix())`.

Reply:
That seems fine to me. The two places this data goes are:
1. Metric emission, in which timestamps are ignored; and
2. `GET /model/costDataModel`, which is practically unused.

I still think it's worth making a sensible decision here, but I don't think that there are any constraints that should make us prefer a different value.

The only small suggestion I might make is that the default value in the final `else` could similarly be zero, to indicate that there was no data at all. But again, I'm not worried about that case at all, really.

Reply:
> The only small suggestion I might make is that the default value in the final else could similarly be zero, to indicate that there was no data at all. But again, I'm not worried about that case at all, really.

Assuming there was full time series data in `[]*util.Vector`, it may be good to set the current timestamp? That way there would be indication at what point allocation could not be calculated due to missing requests and usage data?

```
Value:     0,
Timestamp: float64(time.Now().UTC().Unix()),
```

I'm good either way! As you mentioned this use case is no longer likely to happen.